core/common/src/DateTimePeriod.kt

```diff
@@ -45,83 +52,269 @@ sealed class DateTimePeriod {
-        override val years: Int = 0,
-        override val months: Int = 0,
-        override val days: Int = 0,
-        override val hours: Int = 0,
-        override val minutes: Int = 0,
-        override val seconds: Long = 0,
-        override val nanoseconds: Long = 0
+    internal override val totalMonths: Int = 0,
+    override val days: Int = 0,
+    override val totalNanoseconds: Long = 0,
 ) : DateTimePeriod()
 
+// TODO: these calculations fit in a JS Number. Possible to do an expect/actual here.
+private fun totalMonths(years: Int, months: Int): Int =
+    when (val totalMonths = years.toLong() * 12 + months.toLong()) {
+        in Int.MIN_VALUE..Int.MAX_VALUE -> totalMonths.toInt()
+        else -> throw IllegalArgumentException("The total number of months in $years years and $months months overflows an Int")
+    }
+
+private fun totalNanoseconds(hours: Int, minutes: Int, seconds: Int, nanoseconds: Long): Long {
+    val totalMinutes: Long = hours.toLong() * 60 + minutes
+    // absolute value at most 61 * Int.MAX_VALUE
+    val totalMinutesAsSeconds: Long = totalMinutes * 60
+    // absolute value at most 61 * 60 * Int.MAX_VALUE < 64 * 64 * 2^31 = 2^43
+    val minutesAndNanosecondsAsSeconds: Long = totalMinutesAsSeconds + nanoseconds / NANOS_PER_ONE
+    // absolute value at most 2^43 + 2^63 / 10^9 < 2^43 + 2^34 < 2^44
+    val totalSeconds = minutesAndNanosecondsAsSeconds + seconds
+    // absolute value at most 2^44 + 2^31 < 2^45
+    return try {
+        multiplyAndAdd(totalSeconds, 1_000_000_000, nanoseconds % NANOS_PER_ONE)
+    } catch (e: ArithmeticException) {
+        throw IllegalArgumentException("The total number of nanoseconds in $hours hours, $minutes minutes, $seconds seconds, and $nanoseconds nanoseconds overflows a Long")
+    }
+}
+
+internal fun buildDateTimePeriod(totalMonths: Int = 0, days: Int = 0, totalNanoseconds: Long): DateTimePeriod =
+    if (totalNanoseconds != 0L)
+        DateTimePeriodImpl(totalMonths, days, totalNanoseconds)
+    else
+        DatePeriod(totalMonths, days)
+
 fun DateTimePeriod(
     years: Int = 0,
     months: Int = 0,
     days: Int = 0,
     hours: Int = 0,
     minutes: Int = 0,
-    seconds: Long = 0,
+    seconds: Int = 0,
     nanoseconds: Long = 0
-): DateTimePeriod = if (hours or minutes != 0 || seconds or nanoseconds != 0L)
-    DateTimePeriodImpl(years, months, days, hours, minutes, seconds, nanoseconds)
-else
-    DatePeriod(years, months, days)
+): DateTimePeriod = buildDateTimePeriod(totalMonths(years, months), days,
+    totalNanoseconds(hours, minutes, seconds, nanoseconds))
 
 @OptIn(ExperimentalTime::class)
 fun Duration.toDateTimePeriod(): DateTimePeriod = toComponents { hours, minutes, seconds, nanoseconds ->
-    DateTimePeriod(hours = hours, minutes = minutes, seconds = seconds.toLong(), nanoseconds = nanoseconds.toLong())
+    buildDateTimePeriod(totalNanoseconds = totalNanoseconds(hours, minutes, seconds, nanoseconds.toLong()))
```

Comment:
Perhaps we could now avoid splitting duration to components and just convert its nanosecond value?

Reply:
Thanks!